Fix android audio 0.5 second delay

MikuAuahDark opened the pull request.

kcat commented:
Is there some way to configure or detect the total buffering delay added by Android? OpenAL Soft's default buffer metrics should only have less than 0.1 second of delay, and this change tries to set just 23ms of delay, which is dangerously small. Relying on "invisible" buffers between OpenSL and whatever Android actually uses for output doesn't seem like a great idea if there's no indication of them.

MikuAuahDark commented:
> OpenAL Soft's default buffer metrics should only have less than 0.1 second of delay

It gives me 0.5 second of delay, tested in 2 devices (ARMv7, Android 4.2.2 and Intel Atom Android 5.0).

Well, for detecting, I have no idea. But [does this one help?](https://developer.android.com/ndk/guides/audio/audio-latency.html#output-latency)

kcat commented:
Would help more if I could do that in native C, as I've yet to find a way to call Java functions from a plain old C library that has a set API. All the examples I can find seem to assume I have access to the main application project and can modify it somehow, which isn't true.

MikuAuahDark commented:
Yeah, I know that. Unfortunately there seems no way to get the JavaVM pointer which is very sad.

kcat commented:
Commits f0c8b7f255b17a2fcfcf622a78794cde07f033cd and 67ffdf7a78ae3e48b1fbdd866ad7c2add7e5b6fd attempt to add getting the JavaVM pointer and the system's sample rate. Can you test to see if they work (i.e. it doesn't crash), and if it fixes the lag? If possible, I'd also be interested in a trace log (which now also goes to the android log via ```__android_log_print```).

MikuAuahDark commented:
I tested it with LOVE v0.10.2 with modified libopenal.so with those patches, and it crashes.

The logcat are below

	D/openal  (27757): AL lib: alc_initconfig: Supported backends: opensl, null, wave
	D/openal  (27757): AL lib: ReadALConfig: Loading config /etc/openal/alsoft.conf...
	D/openal  (27757): AL lib: ReadALConfig: Loading config /etc/xdg/alsoft.conf...
	D/openal  (27757): AL lib: GetConfigValue: Key disable-cpu-exts not found
	D/openal  (27757): AL lib: FillCPUCaps: Extensions: -none-
	D/openal  (27757): AL lib: GetConfigValue: Key rt-prio not found
	D/openal  (27757): AL lib: GetConfigValue: Key resampler not found
	D/openal  (27757): AL lib: GetConfigValue: Key trap-al-error not found
	D/openal  (27757): AL lib: GetConfigValue: Key trap-alc-error not found
	D/openal  (27757): AL lib: GetConfigValue: Key reverb/boost not found
	D/openal  (27757): AL lib: GetConfigValue: Key reverb/emulate-eax not found
	D/openal  (27757): AL lib: GetConfigValue: Key drivers not found
	D/openal  (27757): AL lib: alc_initconfig: Initialized backend "opensl"
	D/openal  (27757): AL lib: alc_initconfig: Added "opensl" for playback
	D/openal  (27757): AL lib: alc_initconfig: Initialized backend "null"
	D/openal  (27757): AL lib: alc_initconfig: Initialized backend "wave"
	D/openal  (27757): AL lib: GetConfigValue: Key excludefx not found
	D/openal  (27757): AL lib: GetConfigValue: Key default-reverb not found
	D/openal  (27757): AL lib: GetConfigValue: Key channels not found
	D/openal  (27757): AL lib: GetConfigValue: Key sample-type not found
	D/openal  (27757): AL lib: GetConfigValue: Key format not found
	D/openal  (27757): AL lib: GetConfigValue: Key frequency not found
	D/openal  (27757): AL lib: GetConfigValue: Key periods not found
	D/openal  (27757): AL lib: GetConfigValue: Key period_size not found
	D/openal  (27757): AL lib: GetConfigValue: Key sources not found
	D/openal  (27757): AL lib: GetConfigValue: Key slots not found
	D/openal  (27757): AL lib: GetConfigValue: Key sends not found
	D/openal  (27757): AL lib: GetConfigValue: Key cf_level not found
	D/openal  (27757): AL lib: GetConfigValue: Key midi/internal-synth not found
	D/openal  (27757): AL lib: SSynth_create: Not using internal MIDI synth
	D/openal  (27757): AL lib: alcOpenDevice: Created device 0x5452ef90, "OpenSL"
	V/openal  (27757): AL lib: ALCdevice_IncRef: 0x5452ef90 increasing refcount to 2
	D/openal  (27757): AL lib: UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x2
	D/openal  (27757): AL lib: GetConfigValue: Key hrtf not found
	E/openal  (27757): AL lib: opensl_reset_playback: engine->CreateAudioPlayer: Parameter invalid
	V/openal  (27757): AL lib: ALCdevice_DecRef: 0x5452ef90 decreasing refcount to 1
	D/dalvikvm(27802): Trying to load lib /data/app-lib/org.love2d.android-1/libopenal.so 0x419bd8b0
	D/dalvikvm(27802): Added shared lib /data/app-lib/org.love2d.android-1/libopenal.so 0x419bd8b0
	D/openal  (27802): AL lib: alc_initconfig: Supported backends: opensl, null, wave
	D/openal  (27802): AL lib: ReadALConfig: Loading config /etc/openal/alsoft.conf...
	D/openal  (27802): AL lib: ReadALConfig: Loading config /etc/xdg/alsoft.conf...
	D/openal  (27802): AL lib: GetConfigValue: Key disable-cpu-exts not found
	D/openal  (27802): AL lib: FillCPUCaps: Extensions: -none-
	D/openal  (27802): AL lib: GetConfigValue: Key rt-prio not found
	D/openal  (27802): AL lib: GetConfigValue: Key resampler not found
	D/openal  (27802): AL lib: GetConfigValue: Key trap-al-error not found
	D/openal  (27802): AL lib: GetConfigValue: Key trap-alc-error not found
	D/openal  (27802): AL lib: GetConfigValue: Key reverb/boost not found
	D/openal  (27802): AL lib: GetConfigValue: Key reverb/emulate-eax not found
	D/openal  (27802): AL lib: GetConfigValue: Key drivers not found
	D/openal  (27802): AL lib: alc_initconfig: Initialized backend "opensl"
	D/openal  (27802): AL lib: alc_initconfig: Added "opensl" for playback
	D/openal  (27802): AL lib: alc_initconfig: Initialized backend "null"
	D/openal  (27802): AL lib: alc_initconfig: Initialized backend "wave"
	D/openal  (27802): AL lib: GetConfigValue: Key excludefx not found
	D/openal  (27802): AL lib: GetConfigValue: Key default-reverb not found
	D/openal  (27802): AL lib: GetConfigValue: Key channels not found
	D/openal  (27802): AL lib: GetConfigValue: Key sample-type not found
	D/openal  (27802): AL lib: GetConfigValue: Key format not found
	D/openal  (27802): AL lib: GetConfigValue: Key frequency not found
	D/openal  (27802): AL lib: GetConfigValue: Key periods not found
	D/openal  (27802): AL lib: GetConfigValue: Key period_size not found
	D/openal  (27802): AL lib: GetConfigValue: Key sources not found
	D/openal  (27802): AL lib: GetConfigValue: Key slots not found
	D/openal  (27802): AL lib: GetConfigValue: Key sends not found
	D/openal  (27802): AL lib: GetConfigValue: Key cf_level not found
	D/openal  (27802): AL lib: GetConfigValue: Key midi/internal-synth not found
	D/openal  (27802): AL lib: SSynth_create: Not using internal MIDI synth
	D/openal  (27802): AL lib: alcOpenDevice: Created device 0x5452ef90, "OpenSL"
	V/openal  (27802): AL lib: ALCdevice_IncRef: 0x5452ef90 increasing refcount to 2
	D/openal  (27802): AL lib: UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x2
	D/openal  (27802): AL lib: GetConfigValue: Key hrtf not found
	E/openal  (27802): AL lib: opensl_reset_playback: engine->CreateAudioPlayer: Parameter invalid
	V/openal  (27802): AL lib: ALCdevice_DecRef: 0x5452ef90 decreasing refcount to 1

kcat commented:
It's best to avoid those other patches, or else we won't be able to tell if the changes have fixed the latency problem. The crash doesn't seem to be with OpenAL Soft itself since OpenSL's ```CreateAudioPlayer``` method is cleanly failing during a reset, which would cause ```alcCreateContext``` to both fail and put the device into a disconnected state.

I've pushed another commit which adds a traces to see the sample rate it gets from the system, and should help prevent ```CreateAudioPlayer``` from failing under certain situations.

MikuAuahDark commented:
Sorry, what I mean "with those patches" is 67ffdf7a78ae3e48b1fbdd866ad7c2add7e5b6fd

I'll try it ASAP.

MikuAuahDark commented:
Unfortunately, it still crashes. It seems doesn't like my system sample rate (which is 14Hz, huh?)

The logcat:

	D/dalvikvm(29465): Trying to load lib /data/app-lib/org.love2d.android-1/libopenal.so 0x419c2530
	D/dalvikvm(29465): Added shared lib /data/app-lib/org.love2d.android-1/libopenal.so 0x419c2530
	D/openal  (29465): AL lib: alc_initconfig: Supported backends: opensl, null, wave
	D/openal  (29465): AL lib: ReadALConfig: Loading config /etc/openal/alsoft.conf...
	D/openal  (29465): AL lib: ReadALConfig: Loading config /etc/xdg/alsoft.conf...
	D/openal  (29465): AL lib: GetConfigValue: Key disable-cpu-exts not found
	D/openal  (29465): AL lib: FillCPUCaps: Extensions: -none-
	D/openal  (29465): AL lib: GetConfigValue: Key rt-prio not found
	D/openal  (29465): AL lib: GetConfigValue: Key resampler not found
	D/openal  (29465): AL lib: GetConfigValue: Key trap-al-error not found
	D/openal  (29465): AL lib: GetConfigValue: Key trap-alc-error not found
	D/openal  (29465): AL lib: GetConfigValue: Key reverb/boost not found
	D/openal  (29465): AL lib: GetConfigValue: Key reverb/emulate-eax not found
	D/openal  (29465): AL lib: GetConfigValue: Key drivers not found
	D/openal  (29465): AL lib: alc_initconfig: Initialized backend "opensl"
	D/openal  (29465): AL lib: alc_initconfig: Added "opensl" for playback
	D/openal  (29465): AL lib: alc_initconfig: Initialized backend "null"
	D/openal  (29465): AL lib: alc_initconfig: Initialized backend "wave"
	D/openal  (29465): AL lib: GetConfigValue: Key excludefx not found
	D/openal  (29465): AL lib: GetConfigValue: Key default-reverb not found
	D/openal  (29465): AL lib: GetConfigValue: Key channels not found
	D/openal  (29465): AL lib: GetConfigValue: Key sample-type not found
	D/openal  (29465): AL lib: GetConfigValue: Key format not found
	D/openal  (29465): AL lib: GetConfigValue: Key frequency not found
	D/openal  (29465): AL lib: GetConfigValue: Key periods not found
	D/openal  (29465): AL lib: GetConfigValue: Key period_size not found
	D/openal  (29465): AL lib: GetConfigValue: Key sources not found
	D/openal  (29465): AL lib: GetConfigValue: Key slots not found
	D/openal  (29465): AL lib: GetConfigValue: Key sends not found
	D/openal  (29465): AL lib: GetConfigValue: Key cf_level not found
	D/openal  (29465): AL lib: GetConfigValue: Key midi/internal-synth not found
	D/openal  (29465): AL lib: SSynth_create: Not using internal MIDI synth
	D/openal  (29465): AL lib: alcOpenDevice: Created device 0x54531de0, "OpenSL"
	V/openal  (29465): AL lib: ALCdevice_IncRef: 0x54531de0 increasing refcount to 2
	D/openal  (29465): AL lib: UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x4
	D/openal  (29465): AL lib: GetConfigValue: Key hrtf not found
	D/openal  (29465): AL lib: opensl_reset_playback: Got system sample rate 14hz
	D/openal  (29465): AL lib: UpdateDeviceParams: Post-reset: Stereo, Signed Short, 8000hz, 1024 update size x2
	D/openal  (29465): AL lib: GetConfigValue: Key layout_stereo not found
	D/openal  (29465): AL lib: GetConfigValue: Key layout not found
	D/openal  (29465): AL lib: UpdateDeviceParams: HRTF disabled
	D/openal  (29465): AL lib: UpdateDeviceParams: BS2B disabled
	D/openal  (29465): AL lib: GetConfigValue: Key wide-stereo not found
	V/openal  (29465): AL lib: ALCdevice_IncRef: 0x54531de0 increasing refcount to 3
	V/openal  (29465): AL lib: ALCdevice_DecRef: 0x54531de0 decreasing refcount to 2
	D/openal  (29465): AL lib: alcCreateContext: Created context 0x51255f90
	V/openal  (29465): AL lib: ALCcontext_IncRef: 0x51255f90 increasing refcount to 2
	V/openal  (29465): AL lib: ALCdevice_IncRef: 0x54531de0 increasing refcount to 3
	V/openal  (29465): AL lib: ALCdevice_DecRef: 0x54531de0 decreasing refcount to 2
	V/openal  (29465): AL lib: ALCcontext_IncRef: 0x51255f90 increasing refcount to 3
	V/openal  (29465): AL lib: ALCcontext_DecRef: 0x51255f90 decreasing refcount to 2
	V/openal  (29465): AL lib: ALCcontext_IncRef: 0x51255f90 increasing refcount to 3
	V/openal  (29465): AL lib: ALCcontext_DecRef: 0x51255f90 decreasing refcount to 2
	... last 2 message above repeated about more than 20 times ...

The strange thing is `D/openal  (29465): AL lib: opensl_reset_playback: Got system sample rate 14hz`

kcat commented:
> The strange thing is D/openal (29465): AL lib: opensl_reset_playback: Got system sample rate 14hz

Indeed, that is very strange. I've pushed a couple more commits which adds a bit more logging to the JNIEnv calls, which can hopefully give a bit more insight into what's being passed back and forth between C and Java. Can you test and see what the log prints out now?

As for the crashing, I can't think of anything in OpenAL Soft that would be causing it with these logs. Can you get a backtrace?

MikuAuahDark commented:
Sorry for the late. Still crash at the moment. Looks like my device `PROPERTY_OUTPUT_SAMPLE_RATE` is 0.

Here's the logcat if you're interested

	D/openal  ( 9793): AL lib: alc_initconfig: Supported backends: opensl, null, wave
	D/openal  ( 9793): AL lib: ReadALConfig: Loading config /etc/openal/alsoft.conf...
	D/openal  ( 9793): AL lib: ReadALConfig: Loading config /etc/xdg/alsoft.conf...
	D/openal  ( 9793): AL lib: GetConfigValue: Key disable-cpu-exts not found
	D/openal  ( 9793): AL lib: FillCPUCaps: Extensions: -none-
	D/openal  ( 9793): AL lib: GetConfigValue: Key rt-prio not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key resampler not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key trap-al-error not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key trap-alc-error not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key reverb/boost not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key reverb/emulate-eax not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key drivers not found
	D/openal  ( 9793): AL lib: alc_initconfig: Initialized backend "opensl"
	D/openal  ( 9793): AL lib: alc_initconfig: Added "opensl" for playback
	D/openal  ( 9793): AL lib: alc_initconfig: Initialized backend "null"
	D/openal  ( 9793): AL lib: alc_initconfig: Initialized backend "wave"
	D/openal  ( 9793): AL lib: GetConfigValue: Key excludefx not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key default-reverb not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key channels not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key sample-type not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key format not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key frequency not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key periods not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key period_size not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key sources not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key slots not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key sends not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key cf_level not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key midi/internal-synth not found
	D/openal  ( 9793): AL lib: SSynth_create: Not using internal MIDI synth
	D/openal  ( 9793): AL lib: alcOpenDevice: Created device 0x545e6de0, "OpenSL"
	V/openal  ( 9793): AL lib: ALCdevice_IncRef: 0x545e6de0 increasing refcount to 2
	D/openal  ( 9793): AL lib: UpdateDeviceParams: Pre-reset: Stereo, Float, 44100hz, 1024 update size x4
	D/openal  ( 9793): AL lib: GetConfigValue: Key hrtf not found
	D/openal  ( 9793): AL lib: opensl_reset_playback: Integer: 0x1d400009, parseInt: 0x4bd965f0
	D/openal  ( 9793): AL lib: opensl_reset_playback: Context: 0x1d50000d, AUDIO_SERVICE: 0x4120e088, getSystemService: 0x4be06ea0
	D/openal  ( 9793): AL lib: opensl_reset_playback: AudioManager: 0x1d500011, PROPERTY_OUTPUT_SAMPLE_RATE: 0x4123e2a0, getProperty: 0x0
	D/openal  ( 9793): AL lib: opensl_reset_playback: Context.getSystemService(audio) = 0x1d500015
	D/openal  ( 9793): AL lib: opensl_reset_playback: audMgr.getProperty(android.media.property.OUTPUT_SAMPLE_RATE) = 0x0
	D/openal  ( 9793): AL lib: opensl_reset_playback: Got system sample rate 14hz ((null))
	D/openal  ( 9793): AL lib: UpdateDeviceParams: Post-reset: Stereo, Signed Short, 8000hz, 1024 update size x2
	D/openal  ( 9793): AL lib: GetConfigValue: Key layout_stereo not found
	D/openal  ( 9793): AL lib: GetConfigValue: Key layout not found
	D/openal  ( 9793): AL lib: UpdateDeviceParams: HRTF disabled
	D/openal  ( 9793): AL lib: UpdateDeviceParams: BS2B disabled
	D/openal  ( 9793): AL lib: GetConfigValue: Key wide-stereo not found
	V/openal  ( 9793): AL lib: ALCdevice_IncRef: 0x545e6de0 increasing refcount to 3
	V/openal  ( 9793): AL lib: ALCdevice_DecRef: 0x545e6de0 decreasing refcount to 2
	D/openal  ( 9793): AL lib: alcCreateContext: Created context 0x512feaa0
	V/openal  ( 9793): AL lib: ALCcontext_IncRef: 0x512feaa0 increasing refcount to 2
	V/openal  ( 9793): AL lib: ALCdevice_IncRef: 0x545e6de0 increasing refcount to 3
	V/openal  ( 9793): AL lib: ALCdevice_DecRef: 0x545e6de0 decreasing refcount to 2
	V/openal  ( 9793): AL lib: ALCcontext_IncRef: 0x512feaa0 increasing refcount to 3
	V/openal  ( 9793): AL lib: ALCcontext_DecRef: 0x512feaa0 decreasing refcount to 2

kcat commented:
> ```D/openal  ( 9793): AL lib: opensl_reset_playback: AudioManager: 0x1d500011, PROPERTY_OUTPUT_SAMPLE_RATE: 0x4123e2a0, getProperty: 0x0```

Seems the ```AudioManager.getProperty``` method is being returned as ```NULL```, so calls to it implicitly fail. Interesting that ```Integer.parseInt``` returns ```14``` for null parameter values.

It looks like I had the wrong return type for ```getProperty```, which returns a ```String``` rather than an ```Object```. Hopefully that will get the function call to work properly. Can you pull the latest code and try?

MikuAuahDark commented:
I'll check it out tomorrow.

One additional note: you may want to enable such thing (using optimal audio buffer size) only for Android API Level 17 or higher, otherwise the application will crash calling to null function (`string AudioManager.getProperty(string);` was added in API Level 17, Android 4.2.2)

MikuAuahDark commented:
Testing latest patch again, and unfortunately still crash.

opensl.c opensl_reset_playback: `strchars = JCALL(env,GetStringUTFChars)(srateStr, NULL);` seems return NULL, but strangely it returns 14hz.

The worst part is that I don't have device for testing again. My Android 5.0 phone is not (and without wiping it's data is impossible) to root.

kcat commented:
What does the trace log have?

MikuAuahDark commented:
Same as the last one I posted, except for this one:

    D/openal  ( 9793): AL lib: opensl_reset_playback: AudioManager: 0x1d500011, PROPERTY_OUTPUT_SAMPLE_RATE: 0x4123e2a0, getProperty: 0x0

While previously `getProperty` is 0x0, in this case, `getProperty` is pointing to address somewhere (most likely the `getProperty` function in Java)

kcat commented:
> Isn't that you write ```TRACE("Got system sample rate %uhz (%s)\n", sampleRate, strchars);``` for that? It seems ```strchars``` is NULL.

Yes, it seems because
```AL lib: opensl_reset_playback: audMgr.getProperty(android.media.property.OUTPUT_SAMPLE_RATE) = 0x0```
The getProperty call is returning NULL for the string object for some reason, so when it tries to get the UTF-8 chars, it gets a NULL pointer that the TRACE prints as ```(null)```.

I pushed a change to make ```getSystemService``` be called as a non-static method. I don't know if that will fix (or further break) anything though. If it was returning a bad AudioManager object, that could account for the getProperty call subsequently failing too.

kcat commented:
I've gone back and disabled the use of Java to try and get the sample rate, because it wasn't working. I've unfortunately still not found a way to get the Java ```AudioManager``` from within the library, as it relies on getting a Java Context object that the app has. However, with the Git version you should be able to get the sample rate in your Java code, and pass that to ```alcCreateContext``` with the ```ALC_FREQUENCY``` property. If a mismatched sample rate is cause of the excessive latency, that should help fix it.

MikuAuahDark commented:
Sorry for the very very long response.

********************************

Ok I was experiment with more and more lower buffer size and lowering it as low as 8 bytes, it gives me delay about 100-300ms, which seems to be my phone delay.

Hope that helps.

jmiskovic commented:
I can get the latency on the Android to about 160ms by setting NumUpdates to 1 and UpdateSize to 16 (on my phone 8 causes runtime error). This very similar to audio latency on desktop (linux). It's hard to measure how is 160ms distributed between input lag, openal-soft stack and the rest of audio system.

If I wanted to further reduce the latency in openal-soft on both desktop and Android, where should I look?

MikuAuahDark commented:
In my Ubuntu, I didn't notice any delay at all, but yes I can notice that 160ms delay in my Android phone.

mirh commented:
> 160ms 

Idk anything about OAL.. but isn't that a tad high?
Or I mean, at least my shield can go as down as 16ms, is your phone [supposed](http://superpowered.com/latency) to suck all that much?

MikuAuahDark commented:
Yeah, but I was looking at a game engine which uses almost zero latency audio with OpenSL.

jmiskovic commented:
Just tested with superpowered's latency test. My Xiaomi Redmi 3S gets about 50ms (with large spread, from 32ms to 65ms). Now I should try to build a test app that records latency without input lag, so the comparison with openal-soft is meaningful.

I also re-measured desktop app and it came out as 90ms. My previous measurement incorrect because I was triggering sound with touch-sensitive display, which it seems introduces delay to be able to detect double click.

mirh commented:
Not saying you should expect similar results from a "complex suite" like openal, but on desktop JACK/ASIO should be the golden comparison standard (and iirc I had my xonar/realtek go down to some ms.. Though I didn't measured anything, so that may mean nothing)

Sharabaddin commented:
If getProperty(PROPERTY_OUTPUT_SAMPLE_RATE) return null .You must manually set OUTPUT_SAMPLE_RATE = 44100 (this is most values on Android Device`s.)

kcat commented:
The default sample rate is 44100. Until I can work out how to get the ```AudioManager``` (which it seems will likely need an extension for the app to manually pass it in), you can query the device's sample rate and make OpenAL try to match it:
```
ALCint attrs[] = {
    ALC_FREQUENCY, detected_sample_rate,
    0
};
context = alcCreateContext(device, attrs);
```

One other possibility for the latency might be if OpenAL Soft's update size is different from the device's. I believe I read that the sample rate and update size should match to make the latency as small as possible, though it's not currently possible to set OpenAL's update size from the app. Can you try matching UpdateSize (which is in sample frames) to your device, as well as setting the matching sample rate through alcCreateContext, and let me know if that fixes the latency?

MikuAuahDark commented:
Since the app that uses OpenAL-soft uses SDL2 too, I can retrieve some JNI-related function using SDL2 function by `dlsym` it with `RTLD_DEFAULT`. Unfortunately I lost it because my laptop hard drive died about 2 weeks ago. I'll try my best to recover it if possible.

jmiskovic commented:
One more piece of info - just setting update size to as low value as possible isn't optimal. You should query the audio driver which buffer length (update size) and use the same, just like the sample rate. See [this video](https://youtu.be/d3kfEeMZ65c?t=1703) for more info.

With this technique I was able to get touch-to-sound latency of 65ms, which is as good as it gets for my phone. Right now I have to hard-code sample-rate and update size so my Android app cannot adapt to other devices :(

MikuAuahDark commented:
Ok so I recovered my modified `opensl.c`

```cpp
// ...

#ifdef ANDROID

#include <dlfcn.h>
#include <jni.h>
#include <stdlib.h>

typedef void* jni_get_signature();

void *AndroidGetJNIEnv()
{
	static void* a = NULL;
	static int loaded = 0;
	
	if (loaded == 0)
	{
		void *x = dlsym(RTLD_DEFAULT, "SDL_AndroidGetJNIEnv");
		loaded = 1;
		
		if (x)
			a = ((jni_get_signature*)x)();
	}
	
	return a;
}

jobject AndroidGetActivity()
{
	static jobject activity = NULL;
	static int loaded = 0;
	
	if (loaded == 0)
	{
		void *x = dlsym(RTLD_DEFAULT, "SDL_AndroidGetActivity");
		loaded = 1;
		
		if (x)
			activity = (jobject)((jni_get_signature*)x)();
	}
	
	return activity;
}
#define orxAndroid_GetJNIEnv AndroidGetJNIEnv
#define orxAndroid_GetActivity AndroidGetActivity

/*
The recommended sequence is:

Check for API level 9 or higher, to confirm use of OpenSL ES.
Check for feature "android.hardware.audio.low_latency" using code such as this:
import android.content.pm.PackageManager;
...
PackageManager pm = getContext().getPackageManager();
boolean claimsFeature = pm.hasSystemFeature(PackageManager.FEATURE_AUDIO_LOW_LATENCY);
Check for API level 17 or higher, to confirm use of android.media.AudioManager.getProperty().
Get the native or optimal output sample rate and buffer size for this device's primary output stream, using code such as this:
import android.media.AudioManager;
...
AudioManager am = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
String sampleRate = am.getProperty(AudioManager.PROPERTY_OUTPUT_SAMPLE_RATE));
String framesPerBuffer = am.getProperty(AudioManager.PROPERTY_OUTPUT_FRAMES_PER_BUFFER));
Note that sampleRate and framesPerBuffer are Strings. First check for null and then convert to int using Integer.parseInt().
Now use OpenSL ES to create an AudioPlayer with PCM buffer queue data locator.
*/

int HasFeatureAudioLowLatency() {
    JNIEnv *env;
    jboolean claimsFeature = JNI_FALSE;

    env = orxAndroid_GetJNIEnv();

    if(env != NULL) {
        jclass objClass;
        jfieldID field;

        (*env)->PushLocalFrame(env, 256);

        /* check if running on JELLY_BEAN_MR1 or later */
        objClass = (*env)->FindClass(env, "android/os/Build$VERSION");
        field = (*env)->GetStaticFieldID(env, objClass, "SDK_INT", "I");

        if((*env)->GetStaticIntField(env, objClass, field) >= 17) {
            jclass contextClass, packageManagerClass;
            jstring FEATURE_AUDIO_LOW_LATENCY;
            jmethodID method;
            jobject packageManager;
            jobject context;

            context = orxAndroid_GetActivity();
            contextClass = (*env)->FindClass(env, "android/content/Context");
            method = (*env)->GetMethodID(env, contextClass, "getPackageManager", "()Landroid/content/pm/PackageManager;");
            packageManager = (*env)->CallObjectMethod(env, context, method);

            packageManagerClass = (*env)->FindClass(env, "android/content/pm/PackageManager");
            field = (*env)->GetStaticFieldID(env, packageManagerClass, "FEATURE_AUDIO_LOW_LATENCY", "Ljava/lang/String;");
            FEATURE_AUDIO_LOW_LATENCY = (*env)->GetStaticObjectField(env, packageManagerClass, field);

            method = (*env)->GetMethodID(env, packageManagerClass, "hasSystemFeature", "(Ljava/lang/String;)Z");
            claimsFeature = (*env)->CallBooleanMethod(env, packageManager, method, FEATURE_AUDIO_LOW_LATENCY);
        }

        (*env)->PopLocalFrame(env, NULL);
    }

    return claimsFeature == JNI_TRUE ? 1 : 0;
}

ALuint GetFrequency(ALuint freq)
{
    ALuint result;
    JNIEnv *env;

    result = freq;
    env = orxAndroid_GetJNIEnv();


    if(env != NULL) {
        jclass contextClass, audioManagerClass;
        jfieldID field;
        jstring AUDIO_SERVICE, PROPERTY_OUTPUT_SAMPLE_RATE, sampleRateString;
        jmethodID getSystemService, getProperty;
        jobject audioManager, context;
        const char *sampleRate;

        (*env)->PushLocalFrame(env, 256);
 
        context = orxAndroid_GetActivity();
        contextClass = (*env)->FindClass(env, "android/content/Context");
        field = (*env)->GetStaticFieldID(env, contextClass, "AUDIO_SERVICE", "Ljava/lang/String;");
        AUDIO_SERVICE = (*env)->GetStaticObjectField(env, contextClass, field);
        getSystemService = (*env)->GetMethodID(env, contextClass, "getSystemService", "(Ljava/lang/String;)Ljava/lang/Object;");
        audioManager = (*env)->CallObjectMethod(env, context, getSystemService, AUDIO_SERVICE);
        audioManagerClass = (*env)->FindClass(env, "android/media/AudioManager");
        field = (*env)->GetStaticFieldID(env, audioManagerClass, "PROPERTY_OUTPUT_SAMPLE_RATE", "Ljava/lang/String;");
        getProperty = (*env)->GetMethodID(env, audioManagerClass, "getProperty", "(Ljava/lang/String;)Ljava/lang/String;");
        PROPERTY_OUTPUT_SAMPLE_RATE = (*env)->GetStaticObjectField(env, audioManagerClass, field);
        sampleRateString = (*env)->CallObjectMethod(env, audioManager, getProperty, PROPERTY_OUTPUT_SAMPLE_RATE);

        if(sampleRateString != NULL)
        {
            sampleRate = (*env)->GetStringUTFChars(env, sampleRateString, NULL);
            result = (int)strtol(sampleRate, (char **)NULL, 10);
            (*env)->ReleaseStringUTFChars(env, sampleRateString, sampleRate);
        }

        (*env)->PopLocalFrame(env, NULL);
    }


    return result;
}

ALuint GetBufferSize(ALuint bufferSize)
{
    ALuint result;
    JNIEnv *env;

    result = bufferSize;
    env = orxAndroid_GetJNIEnv();


    if(env != NULL) {
        jclass contextClass, audioManagerClass;
        jfieldID field;
        jstring AUDIO_SERVICE, PROPERTY_OUTPUT_FRAMES_PER_BUFFER, framesPerBufferString;
        jmethodID getSystemService, getProperty;
        jobject audioManager, context;
        const char *framesPerBuffer;

        (*env)->PushLocalFrame(env, 256);
 
        context = orxAndroid_GetActivity();
        contextClass = (*env)->FindClass(env, "android/content/Context");
        field = (*env)->GetStaticFieldID(env, contextClass, "AUDIO_SERVICE", "Ljava/lang/String;");
        AUDIO_SERVICE = (*env)->GetStaticObjectField(env, contextClass, field);
        getSystemService = (*env)->GetMethodID(env, contextClass, "getSystemService", "(Ljava/lang/String;)Ljava/lang/Object;");
        audioManager = (*env)->CallObjectMethod(env, context, getSystemService, AUDIO_SERVICE);
        audioManagerClass = (*env)->FindClass(env, "android/media/AudioManager");
        field = (*env)->GetStaticFieldID(env, audioManagerClass, "PROPERTY_OUTPUT_FRAMES_PER_BUFFER", "Ljava/lang/String;");
        getProperty = (*env)->GetMethodID(env, audioManagerClass, "getProperty", "(Ljava/lang/String;)Ljava/lang/String;");
        PROPERTY_OUTPUT_FRAMES_PER_BUFFER = (*env)->GetStaticObjectField(env, audioManagerClass, field);
        framesPerBufferString = (*env)->CallObjectMethod(env, audioManager, getProperty, PROPERTY_OUTPUT_FRAMES_PER_BUFFER);

        if(framesPerBufferString != NULL)
        {
            framesPerBuffer = (*env)->GetStringUTFChars(env, framesPerBufferString, NULL);
            result = (int)strtol(framesPerBuffer, (char **)NULL, 10);
            (*env)->ReleaseStringUTFChars(env, framesPerBufferString, framesPerBuffer);
        }

        (*env)->PushLocalFrame(env, 256);
    }

    return result;
}

#endif
// ...
static ALCboolean opensl_reset_playback(ALCdevice *Device)
{
    // ...
    
#ifdef ANDROID

    /* try to enable low-latency */
    _wc_freq = Device->Frequency = GetFrequency(Device->Frequency);
    _wc_bufsize = Device->UpdateSize = GetBufferSize(Device->UpdateSize);

#endif
    // ...
}
```

Do you think it will work?

kcat commented:
> Do you think it will work?

It's a start, at least. It'll obviously only work with apps that use SDL and set those properties before setting up an OpenAL context. I imagine many apps will, and it's fine as a fallback, but a proper method to get it from the app would be good too.

Overall, one thing you'll probably want to do is scale the device's ```NumUpdates``` based on how much ```UpdateSize``` and ```Frequency``` changes to match the output, so that the overall latency is about the same as what those values originally should have given (e.g. if ```UpdateSize``` is 1024 and gets changed to 512 to match the device, then ```NumUpdates``` should double).

MikuAuahDark commented:
Unfortunately that code crashed the app. Tried to use `AttachCurrentThread` without success, so there's something wrong with my code above.

kcat commented:
Looks like ```GetBufferSize``` calls ```PushLocalFrame``` twice, whereas ```GetFrequency``` calls ```PushLocalFrame``` and ```PopLocalFrame```. Could that be the cause?

MikuAuahDark commented:
Well, instead of using that code, I fallback to your previous fix in February and modify it a bit to get activity from `SDL_AndroidGetActivity` using `dlsym`. It looks like successfully got the correct sample rate, but the audio latency is still there.

MikuAuahDark commented:
Probably very old, but gonna add this up.

Since you added "FIXME: Disabled until I figure out how to get the Context needed for the getSystemService call." then I think this might solve the issue. https://stackoverflow.com/a/20240883
